src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.Blocking.cs

```diff
@@ -0,0 +1,384 @@
+
+                // After the thread count based on MinThreads is reached, this value (after it is multiplied by the processor
+                // count) specifies how many additional threads may be created without a delay
+                int blocking_threadsToAddWithoutDelay_procCountFactor =
+                    AppContextConfigHelper.GetInt32Config(
+                        "System.Threading.ThreadPool.Blocking.ThreadsToAddWithoutDelay_ProcCountFactor",
+                        1,
+                        false);
+
+                // After the thread count based on MinThreads is reached, this value (after it is multiplied by the processor
+                // count) specifies how many additional threads may be created, perhaps with delays, before fallback. In
+                // fallback mode a thread would be created if starvation is detected and typically with longer delays.
+                int blocking_maxThreadsToAddBeforeFallback_procCountFactor =
+                    AppContextConfigHelper.GetInt32Config(
+                        "System.Threading.ThreadPool.Blocking.MaxThreadsToAddBeforeFallback_ProcCountFactor",
+                        10,
+                        false);
+
+                // After the thread count based on ThreadsToAddWithoutDelay is reached, this value (after it is multiplied by
+                // the processor count) specifies after how many threads an additional DelayStepMs would be added to the delay
+                // before each new thread is created
+                int blocking_threadsPerDelayStep_procCountFactor =
+                    AppContextConfigHelper.GetInt32Config(
+                        "System.Threading.ThreadPool.Blocking.ThreadsPerDelayStep_ProcCountFactor",
+                        1,
+                        false);
+
+                // After the thread count based on ThreadsToAddWithoutDelay is reached, this value specifies how much additional
+                // delay to add per ThreadsPerDelayStep threads, which would be applied before each new thread is created
+                DelayStepMs =
+                    (uint)AppContextConfigHelper.GetInt32Config(
+                        "System.Threading.ThreadPool.Blocking.DelayStepMs",
+                        25,
+                        false);
+
+                // After the thread count based on ThreadsToAddWithoutDelay is reached, this value specifies the max delay to
+                // use before each new thread is created, until fallback. Fallback mode is used after the thread count based on
+                // MaxThreadsToAddBeforeFallback is reached, see that config value for more information.
+                MaxDelayUntilFallbackMs =
+                    (uint)AppContextConfigHelper.GetInt32Config(
+                        "System.Threading.ThreadPool.Blocking.MaxDelayUntilFallbackMs",
+                        250,
+                        false);
```

Comment:
There's a lot of policy in here, and a lot of knobs to go with it.  Do you have a sense for how all of this is going to behave in the real-world, and if/how someone would utilize these knobs effectively?  How did you arrive at this specific set and also the defaults employed?

Reply:
Some of the criteria used:
- Have a good replacement for setting the `MinThreads` as a workaround
  - This would now be to set `ThreadsToAddWithoutDelay` to the equivalent and set `MaxThreadsToAddBeforeFallback` to something higher to give some buffer for spikes that may need more threads
  - `MaxThreadsToAddBeforeFallback` could also be set to a large value to effectively unlimit the heuristic
- Use progressive delays to avoid creating too many threads too quickly
  - Without that, it would be conceivable that `MaxThreadsToAddBeforeFallback` threads would be created in short order to respond to even a short sync-over-async IO before the IO even completes (if there are that many work items that would block on the async work)
  - The delay also helps to create a high watermark of how many threads were necessary last time to unblock, so that when there's a limit to how many work items would block, it would quickly release existing waiting threads to let other work be done meanwhile
  - The larger the number of threads that get unblocked all at once, the higher the latency of their processing would be after unblocking. There's probably not a good solution to this.
  - Ideally it not require as many threads for an async IO completion to unblock waiting threads, sort of like on Windows where a separate pool of threads handles IO completions, needs some experimentation
  - It's not always clear that adding more threads would help, more so in starvation-type cases
- No one set of defaults will work well for all cases, use conservative defaults to start with
  - The current defaults are much more agressive than before
  - The `MaxThreadsToAddBeforeFallback_ProcCountFactor` of `10` came from a prior discussion where we felt that adding 10x the proc count relatively quickly may not be too bad
  - The defaults can be made more aggressive easily, but it would be difficult to make the defaults more conservative since apps that work well with the defaults may not after that without configuration
  - The really bad cases where many 100s or even 1000s of threads need to be created will likely need to configure for the app's situation based on expected workload and how bad it can get, in order to work around the issue
- Make things sufficiently configurable
  - It would have been nice to make configurable the delay threshold for detecting starvation and the delay used to add threads during continuous starvation. Now, for sync-over-async they delay and rate of progression in delays can be adjusted.
  - Similarly to hill climbing config values, the config values don't have to be used but it can be helpful to enable the freedom to configure them
  - I expect I would suggest most users running into bad blocking due to sync-over-async to configure `ThreadsToAddWithoutDelay` and `MaxThreadsToAddBeforeFallback`, and perhaps `MaxDelayUntilFallbackMs` to control the thread injection latency for spikes
- I intend to use the same config values (maybe with a couple of others) for improving the starvation heuristic similarly in the future
  - Starvation is a bit different and may need a few quirks, but hopefully we can use something similar

Reply:
Decided to remove `MaxThreadsToAddBeforeFallback` and renamed `MaxDelayBeforeFallbackMs` to `MaxDelayMs` in the latest commit. The max threads limit before falling back to starvation seems unnecessary, it would be unlimited for starvation anyway. Now the only time it would fall back to starvation is in low-memory situations.